File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.js
function solution(ingredient) {
    
    var stack = [];
    var answer = 0;
    
    for(let i=0; i<ingredient.length; i++){
        stack.push(ingredient[i]);
        var len = stack.length;
        if(len >= 4){
            if(stack[len-1] == 1 && stack[len-2] == 3 && stack[len-3] == 2 && stack[len-4] == 1){
                answer++;
                for(let j=0; j<4; j++){
                    stack.pop();
                }
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] ingredient) {
        
        
        Deque<Integer> que = new ArrayDeque<>();
        int answer = 0;
        
        for(int i=0; i<ingredient.length; i++){
            if(ingredient[i] == 1){
                if(!que.isEmpty() && que.peekLast() == 3){
                    que.pollLast();
                }else{
                    que.add(1);
                    continue;
                }
                if(!que.isEmpty() && que.peekLast() == 2){
                    que.pollLast();
                }else{
                    que.add(3);
                    que.add(1);
                    continue;
                }
                if(!que.isEmpty() && que.peekLast() == 1){
                    que.pollLast();
                    answer++;
                }else{
                    que.add(2);
                    que.add(3);
                    que.add(1);
                    continue;
                }
            }else{
                que.add(ingredient[i]);
            }
        }

        return answer;
    }
}